File: packages/contracts-typescript/src/wasm-artifact.ts
// Packaged wasm artifact resolution + pinned-sha guard (WS-G, M3 P3CORPUS).
//
// The single CEL engine is a reproducible relay_cel_wasm.wasm produced by the
// packages/cel-wasm/conformance/build.sh deterministic recipe. WS-G ships that
// artifact as PACKAGE DATA of @epochly/relay-contracts (under src/_wasm/) so an
// INSTALLED package can resolve the engine WITHOUT the (gitignored)
// crate/target/ tree. This is the TypeScript mirror of the Python
// packages/contracts/src/relay_contracts/wasm_artifact.py, shipping the SAME
// bytes (the SAME pinned sha) for the npm ecosystem.
//
// This module is the single TS source of truth for:
//
//   - WASM_PACKAGE_DATA_SUBPATH -- the `exports` subpath key
//     ("./wasm") under which @epochly/relay-contracts publishes the wasm, so a
//     consumer resolves it via require.resolve('@epochly/relay-contracts/wasm')
//     or import.meta.resolve of the same subpath -- NOT a crate/target/
//     relative path.
//   - WASM_PACKAGE_DATA_RELPATH -- the wasm's path relative to the package root
//     (src/_wasm/relay_cel_wasm.wasm). Ships in `files` (the `src` dir is in
//     the published `files` list) and is the `exports["./wasm"]` target.
//   - WASM_PINNED_SHA256 -- the full sha256 of the build.sh deterministic-
//     recipe artifact (the [repro] PASS sha). Equal to the Python
//     WASM_PINNED_SHA256 (one shared truth across ecosystems). The shipped
//     package-data wasm MUST hash to this value; a guard test fails on a
//     tampered / stale vendored artifact.
//   - resolvePackagedWasmPath() -- resolve the package-data wasm to a concrete
//     on-disk path. Non-throwing: returns null for an absent artifact so the
//     caller maps a missing artifact to a structured engine error rather than
//     letting a bare ENOENT escape.
//
// This is NOT signing. M3 PINS the sha (a checked-in constant + an on-disk-hash
// guard); the wasm is Apache-2.0-portable CODE, not trust-anchor key material.
// Signing the artifact (KMS / transparency log) lives in relay-platform and is
// explicitly out of scope here (CLAUDE.md banned pattern #14).
//
// ASCII-only per CLAUDE.md "ASCII-Safe Source".

import { existsSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

// The `exports` subpath KEY under which the package publishes the wasm. A
// consumer resolves the artifact via require.resolve / import.meta.resolve of
// `@epochly/relay-contracts/wasm` (package name + this subpath minus the leading
// dot), so the wasm is locatable from an INSTALLED package without a
// crate/target/ relative path. package.json `exports` maps this key to
// WASM_PACKAGE_DATA_RELPATH.
export const WASM_PACKAGE_DATA_SUBPATH = "./wasm" as const;

// The vendored wasm's path RELATIVE TO the package root. Ships in the published
// tarball because the `src` directory is in package.json `files`. This is the
// `exports["./wasm"]` target and the path resolvePackagedWasmPath() returns.
// POSIX-style separator; resolved per-platform via node:path.
export const WASM_PACKAGE_DATA_RELPATH =
  "src/_wasm/relay_cel_wasm.wasm" as const;

// The vendored `.mjs` LOADER's path RELATIVE TO the package root. The canonical
// loader (packages/cel-wasm/typescript/relay-cel-wasm.mjs) is a REPO sibling
// that does NOT ship in this package's `files`, so an installed
// @epochly/relay-contracts could resolve the .wasm (via WASM_PACKAGE_DATA_RELPATH)
// yet NOT construct the wasm backend -- the loader was missing from every
// packaged path. WS-G ships a git-tracked VENDORED COPY of the canonical loader
// here (src/_wasm/relay-cel-wasm.mjs, in package.json `files` via the `src` dir),
// so a fresh-installed package can LOAD the wasm, not only locate it. This is the
// TypeScript mirror of the Python WASM_LOADER_PACKAGE_DATA_RELPATH
// (packages/contracts/src/relay_contracts/wasm_artifact.py). Because the copy is
// a git-tracked DUPLICATE of the canonical source, a BYTE-IDENTITY drift guard
// (wasm_loader_package_data.test.ts) fails CI if the two diverge -- no silent
// drift. POSIX-style separator; resolved per-platform via node:path.
export const WASM_LOADER_PACKAGE_DATA_RELPATH =
  "src/_wasm/relay-cel-wasm.mjs" as const;

// The full sha256 of the reproducible build.sh deterministic-recipe artifact
// (the `[repro] PASS: byte-deterministic (<sha>)` value). EQUAL to the Python
// WASM_PINNED_SHA256 (packages/contracts/src/relay_contracts/wasm_artifact.py):
// both ecosystems ship the SAME bytes. The shipped package-data wasm MUST hash
// to this. PINNED, not signed (see module docstring).
export const WASM_PINNED_SHA256 =
  "431d966b2818ef4539a4f6b78e2903a4d6911c6b6352e256e35531a44f992511" as const;

// The package root, anchored to THIS module's on-disk location. wasm-artifact.ts
// lives at packages/contracts-typescript/src/, so the package root is one level
// up. Resolution is anchored to the imported module, not a cwd-relative path, so
// it works from the dev tree and from an installed package alike (the `src` dir
// ships in `files`, so src/_wasm/relay_cel_wasm.wasm exists in both layouts).
function packageRoot(): string {
  const here = dirname(fileURLToPath(import.meta.url));
  return resolve(here, "..");
}

// Resolve the package-data wasm to a concrete on-disk path, gated on presence.
//
// When `override` is supplied it is treated as an explicit candidate path
// (used by the presence-gate test to point the resolver at a specific --
// possibly absent -- artifact). Otherwise the package-data path
// (WASM_PACKAGE_DATA_RELPATH under the package root) is used.
//
// Non-throwing: returns the resolved path when it exists as a regular file, else
// null, so the caller can map a missing artifact to a structured engine error
// (RelayCelEngineError / RELAY-CEL-009) rather than letting a bare ENOENT
// escape. Never throws for an absent artifact.
export function resolvePackagedWasmPath(override?: string): string | null {
  const candidate =
    override ?? resolve(packageRoot(), WASM_PACKAGE_DATA_RELPATH);
  try {
    if (!existsSync(candidate)) {
      return null;
    }
    if (!statSync(candidate).isFile()) {
      return null;
    }
  } catch {
    // Any filesystem error (permissions, broken symlink, etc.) is treated as
    // "not resolvable" -- the caller maps that to a structured engine error.
    return null;
  }
  return candidate;
}

// Resolve the package-data `.mjs` LOADER to a concrete on-disk path, gated on
// presence. Mirrors resolvePackagedWasmPath() for the loader source: resolution
// is anchored to THIS module's on-disk location (so it works from the dev tree
// and an installed package alike -- the `src` dir ships in `files`), NOT a cwd-
// or crate/target/-relative path. The TS counterpart of the Python
// resolve_packaged_wasm_loader_path (wasm_artifact.py).
//
// When `override` is supplied it is treated as an explicit candidate path (used
// by the absent-loader test). Otherwise the package-data path
// (WASM_LOADER_PACKAGE_DATA_RELPATH under the package root) is used.
//
// Non-throwing: returns the resolved path when it exists as a regular file, else
// null, so the caller can fall back to the repo sibling (dev tree) or map a
// missing loader to a structured engine error rather than letting a bare ENOENT
// escape. Never throws for an absent loader.
export function resolvePackagedLoaderPath(override?: string): string | null {
  const candidate =
    override ?? resolve(packageRoot(), WASM_LOADER_PACKAGE_DATA_RELPATH);
  try {
    if (!existsSync(candidate)) {
      return null;
    }
    if (!statSync(candidate).isFile()) {
      return null;
    }
  } catch {
    // Any filesystem error (permissions, broken symlink, etc.) is treated as
    // "not resolvable" -- the caller falls back / maps to a structured error.
    return null;
  }
  return candidate;
}

// Resolve the wasm artifact path to pass to the `.mjs` RelayCel loader, with the
// EXPLICIT precedence (ROBOREV round-2 finding G):
//
//   1. an explicit `wasmPath` (the caller-supplied override) -- highest;
//   2. `process.env.CEL_WASM` (the operator's artifact override);
//   3. the WS-G PACKAGE-DATA wasm (resolvePackagedWasmPath()) -- lowest;
//   4. `undefined` -- no host-resolved path, so the loader applies its OWN
//      default (its self-relative package-data probe / crate-target build).
//
// The bug: when no explicit `wasmPath` was given and the packaged wasm resolved,
// the host passed the PACKAGE-DATA path straight to the loader, so the loader's
// own `wasmPath || process.env.CEL_WASM || defaultWasmPath()` chain never saw
// CEL_WASM -- the operator's override was silently ignored while the docs still
// claimed it was honored. This resolver makes CEL_WASM take precedence over the
// packaged wasm: the packaged wasm is returned ONLY when neither an explicit path
// nor CEL_WASM is set.
//
// CEL_WASM is the wasm-ARTIFACT-PATH env var (NOT the RELAY_CEL_ENGINE engine
// SELECTOR), so reading it here does not affect engine-selection determinism (the
// default-engine guard scans only for RELAY_CEL_ENGINE and explicitly exempts
// CEL_WASM). An EMPTY/whitespace CEL_WASM is treated as UNSET so a blank
// assignment does not point the loader at "" (which would fail to load).
//
// Returns a concrete path string, or `undefined` to defer to the loader's own
// default. Never throws.
export function resolveWasmPathForLoader(
  wasmPath?: string,
): string | undefined {
  if (wasmPath !== undefined) {
    return wasmPath;
  }
  const envPath = process.env.CEL_WASM;
  if (envPath !== undefined && envPath.trim() !== "") {
    return envPath;
  }
  const packaged = resolvePackagedWasmPath();
  if (packaged !== null) {
    return packaged;
  }
  return undefined;
}

File: packages/contracts/src/relay_contracts/wasm_artifact.py
"""Packaged wasm artifact resolution + pinned-sha guard (WS-G, M3 P3CORPUS).

The single CEL engine is a reproducible ``relay_cel_wasm.wasm`` produced by the
``packages/cel-wasm/conformance/build.sh`` deterministic recipe. WS-G ships that
artifact as PACKAGE DATA of ``relay_contracts`` so a fresh-installed wheel can
load the engine via :func:`importlib.resources.files` WITHOUT the (gitignored)
``crate/target/`` tree present.

This module is the single source of truth for:

  - :data:`WASM_PACKAGE_DATA_RELPATH` -- the data path of the vendored wasm
    relative to the imported ``relay_contracts`` package root, resolvable via
    ``importlib.resources.files('relay_contracts').joinpath(...)``;
  - :data:`WASM_PINNED_SHA256` -- the full sha256 of the build.sh
    deterministic-recipe artifact (the ``[repro] PASS`` sha). The shipped
    package-data wasm MUST hash to this value; a guard test
    (``test_wasm_pinned_sha_matches_packaged_artifact_on_disk``) fails on a
    tampered / stale vendored artifact;
  - :func:`resolve_packaged_wasm_path` -- resolve the package-data wasm to a
    concrete on-disk path (the wheel layout keeps it on the filesystem;
    ``importlib.resources.as_file`` would be required only for a zipped wheel,
    which Relay does not ship for this package).

This is NOT signing. M3 PINS the sha (a checked-in constant + an on-disk-hash
guard); the wasm is Apache-2.0-portable CODE, not trust-anchor key material.
Signing the artifact (KMS / transparency log) lives in ``relay-platform`` and is
explicitly out of scope here (CLAUDE.md banned pattern #14).

ASCII-only per CLAUDE.md "ASCII-Safe Source".
"""
from __future__ import annotations

import hashlib
import importlib.resources
from pathlib import Path

# The vendored wasm's data path RELATIVE TO the imported ``relay_contracts``
# package root. Resolved via ``importlib.resources.files('relay_contracts')``,
# so it works from an installed wheel (no crate/target/ dependency). POSIX-style
# separator; ``importlib.resources`` joinpath normalizes per-platform.
WASM_PACKAGE_DATA_RELPATH: str = "_wasm/relay_cel_wasm.wasm"

# The vendored wasm LOADER module's data path RELATIVE TO the imported
# ``relay_contracts`` package root. The loader
# (``packages/cel-wasm/python/relay_cel_wasm.py``) is a loose module with no
# pyproject, so it is NOT importable as a top-level package in a wheel-only
# install. WS-G ships a git-tracked VENDORED COPY of the canonical loader here
# (``src/relay_contracts/_wasm/relay_cel_wasm.py``) and force-includes it into
# the wheel (``[tool.hatch.build.targets.wheel.force-include]``, the SAME
# mechanism + the SAME vendored-copy pattern the ``.wasm`` uses -- a git-tracked
# copy in ``src/_wasm/``), so a fresh-installed wheel can LOAD the wasm, not only
# locate it. Because the copy is a git-tracked DUPLICATE of the canonical
# source, a BYTE-IDENTITY drift guard
# (``test_wasm_loader_package_data.test_wasm_loader_vendored_copy_is_byte_identical_to_canonical``)
# fails CI if the two diverge -- no silent drift.
WASM_LOADER_PACKAGE_DATA_RELPATH: str = "_wasm/relay_cel_wasm.py"

# The full sha256 of the reproducible build.sh deterministic-recipe artifact
# (the ``[repro] PASS: byte-deterministic (<sha>)`` value). The shipped
# package-data wasm MUST hash to this. PINNED, not signed (see module docstring).
WASM_PINNED_SHA256: str = (
    "431d966b2818ef4539a4f6b78e2903a4d6911c6b6352e256e35531a44f992511"
)


def resolve_packaged_wasm_path() -> Path | None:
    """Resolve the package-data wasm to a concrete on-disk path.

    Uses ``importlib.resources.files('relay_contracts')`` so resolution is
    anchored to the IMPORTED package root, not a source-tree-relative path --
    a fresh-installed wheel locates the wasm the same way the dev tree does.

    Returns the :class:`pathlib.Path` to the wasm when it exists as a regular
    file, else ``None`` (so the caller can map a missing artifact to a
    structured engine error rather than letting a bare ``FileNotFoundError``
    escape). Never raises for an absent artifact.
    """
    try:
        resource = importlib.resources.files("relay_contracts").joinpath(
            WASM_PACKAGE_DATA_RELPATH
        )
    except (ModuleNotFoundError, FileNotFoundError):
        return None
    # ``Traversable.is_file`` works for both filesystem and zip-backed
    # resources; the Relay wheel is a regular (non-zipped) install, so the
    # resource is a real path. Guard against a zipped install by falling back
    # to ``None`` when the resource is not materializable as a filesystem path.
    try:
        if not resource.is_file():
            return None
        # ``importlib.resources.files`` returns a concrete path for a normal
        # (unzipped) wheel; ``Path(str(...))`` materializes it. For the rare
        # zipped-wheel case ``str(resource)`` is not a real path and the file
        # check below catches it.
        path = Path(str(resource))
    except (FileNotFoundError, OSError):
        return None
    if not path.is_file():
        return None
    return path


def resolve_packaged_wasm_loader_path() -> Path | None:
    """Resolve the package-data wasm LOADER module to a concrete on-disk path.

    Mirrors :func:`resolve_packaged_wasm_path` for the loader source: resolution
    is anchored to the IMPORTED ``relay_contracts`` package root via
    ``importlib.resources.files('relay_contracts')``, so a fresh-installed wheel
    locates the loader the same way it locates the ``.wasm`` -- WITHOUT the
    in-repo ``packages/cel-wasm/python`` tree present.

    Returns the :class:`pathlib.Path` to the loader ``.py`` when it exists as a
    regular file, else ``None`` (so the caller maps a missing loader to a
    structured engine error rather than letting a bare ``FileNotFoundError`` /
    ``ImportError`` escape). Never raises for an absent loader.
    """
    try:
        resource = importlib.resources.files("relay_contracts").joinpath(
            WASM_LOADER_PACKAGE_DATA_RELPATH
        )
    except (ModuleNotFoundError, FileNotFoundError):
        return None
    try:
        if not resource.is_file():
            return None
        path = Path(str(resource))
    except (FileNotFoundError, OSError):
        return None
    if not path.is_file():
        return None
    return path


def sha256_of_path(path: Path) -> str:
    """Return the hex sha256 of the bytes at ``path``."""
    return hashlib.sha256(path.read_bytes()).hexdigest()


__all__ = [
    "WASM_LOADER_PACKAGE_DATA_RELPATH",
    "WASM_PACKAGE_DATA_RELPATH",
    "WASM_PINNED_SHA256",
    "resolve_packaged_wasm_loader_path",
    "resolve_packaged_wasm_path",
    "sha256_of_path",
]
